File: CPP/Layer.cpp
#pragma once
#include"Layer.h"



Layer::Layer(int32_t height, int32_t width) 
{
	this->height = height;
	this->width = width;
	this->pixelMap.resize(this->height, std::vector<Pixel>(this->width));

	std::vector<std::vector<Pixel>>::iterator row;
	std::vector<Pixel>::iterator col;


	for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		for (col = row->begin(); col != row->end(); ++col)
			*col = Pixel();
}


Layer::Layer(std::string sourceFileName,int32_t height, int32_t width, uint16_t bitsPerPixel, std::vector<uint8_t>& pixels)
{
	this->sourceFileName = sourceFileName;
	this->height = height;
	this->width = width;

	std::vector<uint8_t>::iterator it;

	/*for (it = pixels.begin(); it!= pixels.end(); ++it) 
	{
		for (int i = 0; i < 4; i++)
			std::cout << *it;

		system("pause");
	}*/

	
	pixelMap.resize(this->height, std::vector<Pixel>(this->width));

	int index = 0;

	std::vector<std::vector<Pixel>>::iterator row;
	std::vector<Pixel>::iterator col;

	if (bitsPerPixel == 32) 
	{
		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index], pixels[index + 1],
					pixels[index + 2], pixels[index + 3]);
				index += 4;
			}
		}
	}
	else if (bitsPerPixel == 24)
	{
		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index], pixels[index + 1],
					pixels[index + 2], 255);
				index += 3;
			}
		}
	}

}
Layer::Layer(std::string sourceFileName,int32_t height, int32_t width, std::vector<uint8_t>& pixels, std::string type)
{
	this->sourceFileName = sourceFileName;
	this->height = height;
	this->width = width;

	this->pixelMap.resize(this->height, std::vector<Pixel>(this->width));

	int index = 0;

	std::vector<std::vector<Pixel>>::iterator row;
	std::vector<Pixel>::iterator col;

	if (strcmp(type.c_str(), "BLACKANDWHITE") == 0)
	{
		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index]<<8, pixels[index]<<8,
					pixels[index]<<8, 255);
				index ++;
			}

		}
	}
	else if (strcmp(type.c_str(), "GRAYSCALE") == 0)
	{

		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index], pixels[index],
					pixels[index], 255);
				index++;
			}
		}
	}
	else if (strcmp(type.c_str(), "RGB") == 0)
	{

		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index], pixels[index + 1],
					pixels[index + 2], 255);
				index += 3;
			}
		}
		
	}
	else if (strcmp(type.c_str(), "BLACKANDWHITE_ALPHA") == 0)
	{
		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index] << 8, pixels[index] << 8,
					pixels[index] << 8, pixels[index + 1] << 8);
				index += 2;
			}


		}
	}
	else if (strcmp(type.c_str(), "GRAYSCALE_ALPHA") == 0)
	{

		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index], pixels[index],
					pixels[index], pixels[index + 1]);
				index += 2;
			}

		}
	}
	else if (strcmp(type.c_str(), "RGB_ALPHA") == 0) 
	{
		for (row = pixelMap.begin(); row != pixelMap.end(); ++row)
		{
			for (col = row->begin(); col != row->end(); ++col)

			{
				*col = Pixel(pixels[index], pixels[index + 1],
					pixels[index + 2], pixels[index + 3]);
				index += 4;
			}
		}
	}

}

void Layer::resize(uint32_t new_height, uint32_t new_width)
{
	std::vector<std::vector<Pixel>> temp (this->height, std::vector<Pixel>(this->width));
	temp = this->pixelMap;

	int32_t old_height = this->height;
	int32_t old_width = this->width;

	pixelMap.clear();
	
	this->height = new_height;
	this->width = new_width;
	this->pixelMap.resize(this->height, std::vector<Pixel>(this->width));

	std::vector<std::vector<Pixel>>::iterator row;
	std::vector<Pixel>::iterator col;

	int32_t i = 0;
	int32_t j = 0;

	for ( row = pixelMap.begin(); row != pixelMap.end(); ++row)
	{
		for (col = row->begin(); col != row->end(); ++col) 
		{
			if (j <= (old_width - 1) && i <= (old_height - 1)) 
				*col = temp[i][j];
			else 
				*col = Pixel();
			j++;
		}
		j = 0;
		i++;
	}
	temp.clear();
}

int32_t Layer::getHeight() const { return this->height; }
int32_t Layer::getWidth() const { return this->width; }
std::vector<std::vector<Pixel>>& Layer::getPixels()  { return this->pixelMap; }

bool Layer::getActivity() const { return this->active; }
void Layer::setActivity(bool newState) { this->active = newState; }

void Layer::setSelection( Selection* s) 
{
	
	this->selection = s;
	this->hasSelection = true;
}
bool Layer::selectedState() const { return this->hasSelection; }
void Layer::fillSelection(int red, int green, int blue)
{
	//podesu granice iteracije
	for (auto r : this->selection->getRechtangles()) 
	{
		
		for (int i = r.getY(); i > (r.getY()- r.getHeight())-1; i--)
		{
			for (int j = r.getX(); j < (r.getX() + r.getWidth()); j++)
			{
				this->pixelMap[i][j].setRed(red);
				this->pixelMap[i][j].setGreen(green);
				this->pixelMap[i][j].setBlue(blue);
			}

		}
	}
		
		
}
Selection* Layer::getSelection()  { return this->selection; }

std::string Layer::getSource() { return this->sourceFileName; }

File: JAVA/Layer.java
package Picture;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Vector;

public class Layer {

     Vector<Vector<Pixel>> pixelMap ;
    private int height;
    private int width;
    private boolean active = true;
    private boolean hasSelection = false;
    private String sourceFile;
    Selection selection=null;


    public Layer(int _height, int _width)
    {
        this.height=_height;
        this.width=_width;

        this.pixelMap= new Vector<Vector<Pixel>>(this.height);

        for(int i =0; i<this.height;i++) {

            Vector<Pixel> row = new Vector<Pixel>(this.width);

            for (int j = 0; j < this.width; j++) {

                row.add(new Pixel());

            }

            pixelMap.add(row);

        }
    }

    public Layer(String _sourceFile, int _height, int _width, ArrayList<Short> pixels) //metoda za bmp slike
    {
        this.sourceFile=_sourceFile;
        this.height=_height;
        this.width=_width;
        this.pixelMap= new Vector<>(this.height);
        int index =0;

       // if(bitsPerPixel==32)
       // {


            for (int i =0; i< this.height;i++) {

                Vector<Pixel> row = new Vector<>(this.width);


                for ( int j =0; j< this.width; j++) {


                    row.add(new Pixel(pixels.get(index), pixels.get(index+1), pixels.get(index+2), pixels.get(index+3)));
                    index+=4;

                }
                pixelMap.add(row);
            }
      //  }

//        if(bitsPerPixel==24)
//        {
//            for (int i =0; i< this.height;i++) {
//
//                Vector<Pixel> row = new Vector<>(this.width);
//
//
//                for ( int j =0; j< this.width; j++) {
//
//                    row.add(new Pixel(pixels.get(index), pixels.get(index+1), pixels.get(index+2), (short)255));
//                    index+=3;
//
//                }
//                pixelMap.add(row);
//            }
//        }
    }

    public Layer(String _sourceFile,int _height, int _width, ArrayList<Short> pxs, String type)
    {
        this.sourceFile=_sourceFile;
        this.height=_height;
        this.width=_width;
        this.pixelMap= new Vector<>(this.height);
        int index =0;

        if (type.equals("BLACKANDWHITE"))
        {
            for (int i =0;i<this.height;i++)
            {
                Vector<Pixel> row = new Vector<Pixel>(this.width);

                for (int j = 0; j<this.width;j++)
                {
                    //row.add(new Pixel(pxs.get(index)<<8, pxs.get(index)<<8, pxs.get(index)<<8, (short) 255));
                    index++;
                }
                this.pixelMap.add(row);
            }
        }
       else if (type.equals("GRAYSCALE"))
       {
           for (int i =0;i<this.height;i++)
           {
               Vector<Pixel> row = new Vector<Pixel>(this.width);

               for (int j = 0; j<this.width;j++)
               {
                   row.add(new Pixel(pxs.get(index),pxs.get(index),pxs.get(index),(short) 255));
                   index++;
               }
               this.pixelMap.add(row);
           }
       }
       else if (type.equals("RGB"))
       {
           for (int i =0;i<this.height;i++)
           {
               Vector<Pixel> row = new Vector<Pixel>(this.width);

               for (int j = 0; j<this.width;j++)
               {
                   row.add(new Pixel(pxs.get(index),pxs.get(index+1),pxs.get(index+2),(short) 255));
                   index+=3;
               }
               this.pixelMap.add(row);
           }
       }
       else if (type.equals("BLACKANDWHITE_ALPHA"))
       {
           for (int i =0;i<this.height;i++)
           {
               Vector<Pixel> row = new Vector<Pixel>(this.width);

               for (int j = 0; j<this.width;j++)
               {
                    //row.add(new Pixel(pxs.get(index)<<8,pxs.get(index)<<8,pxs.get(index)<<8, pxs.get(index+1)<<8));
                    index+=2;

               }
               this.pixelMap.add(row);
           }
       }
       else if (type.equals("GRAYSCALE_ALPHA"))
       {
           for (int i =0;i<this.height;i++)
           {
               Vector<Pixel> row = new Vector<Pixel>(this.width);

               for (int j = 0; j<this.width;j++)
               {
                   row.add(new Pixel(pxs.get(index),pxs.get(index),pxs.get(index),pxs.get(index+1)));
                   index+=2;
               }
               this.pixelMap.add(row);
           }
       }
       else if (type.equals("RGB_ALPHA"))
       {
           for (int i =0;i<this.height;i++)
           {
               Vector<Pixel> row = new Vector<Pixel>(this.width);

               for (int j = 0; j<this.width;j++)
               {
                   row.add(new Pixel(pxs.get(index),pxs.get(index+1),pxs.get(index+2),pxs.get(index+3)));
                   index+=4;
               }
               this.pixelMap.add(row);
           }
       }
    }

    public String getSourceFile() {return  this.sourceFile;}

    public void resize (int newHeight, int newWidth)
    {
        Vector<Vector<Pixel>> newPixelMap = new Vector<>(newHeight);

        int oldHeight = this.height;
        int oldWidth = this.width;

        this.height=newHeight;
        this.width=newWidth;

        for (int i =0;i<newHeight;i++) {

            Vector<Pixel> row = new Vector<>(newWidth);
            for (int j = 0; j < newWidth; j++) {

                if ((j <= (oldWidth - 1)) && (i <= (oldHeight - 1)))
                    row.add(pixelMap.get(i).get(j));
                else
                    row.add(new Pixel());
            }
            newPixelMap.add(row);
        }

        this.pixelMap.clear();
        this.pixelMap=newPixelMap;

    }

    public int getHeight () {return  this.height;}

    public int getWidth() { return this.width; }

    public boolean isActive() { return  this.active; }

    public void setActivity(boolean newState) { this.active = newState; }

    public  boolean getActivity() {return this.active;}

    public void setSelection(Selection s) {this.selection=s;}

    public boolean selectionState() {return  this.hasSelection;}

    public Selection getSelection() {return  this.selection;}

    public void fillSelection (short red, short green, short blue )
    {


        for (Rechtangle r : this.selection.getRechtangles()) {

            for (int i = r.getY(); i < (r.getY()+r.getHeight());i++)
                for (int j = r.getX(); j < (r.getX()+r.getWidth()); j++) {

                    this.pixelMap.get(i).get(j).setRed(red);
                    this.pixelMap.get(i).get(j).setGreen(green);
                    this.pixelMap.get(i).get(j).setBlue(blue);

                }
        }
    }

    public Vector<Vector<Pixel>> getPixelMap() {

        Vector<Vector<Pixel>> copyMap = new Vector<>(this.height);

        for (Vector<Pixel> row: pixelMap) {

            Vector<Pixel> copy = new Vector<>(this.width);

            for (Pixel p: row) {
                copy.add( new Pixel(p));
            }
            copyMap.add(copy);

        }

        //vracace kopiju originalne mape
        return copyMap;
    }



    public void setAlpha(short alpha)
    {
        for (Vector<Pixel> rows: pixelMap) {

            for (Pixel p: rows) {

                p.setAlpha(alpha);
            }

        }
    }

}

